### PR TITLE
Discover and install repositories that provide tools required by imported workflows.

### PR DESCRIPTION
**Pull request by davebx**
Repositories can be previewed and installed or added to the queue for later installation.
- [x] Need to handle cases where the workflow requires tools from multiple repositories.
- [x] May want to handle cases where a workflow uses tools from multiple toolsheds.
- [x] Pop up a modal with the tool form when clicking on the tool button.

Screenshot

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 1]
```diff
@@ -65,8 +68,35 @@ def index(self, trans, **kwd):
...
+    def __is_missing( self, trans, tool_id ):
+        # First make sure the tool ID A) Is valid, and B) Corresponds to a
+        # toolshed tool.
+        if tool_id is None:
+            return False
+        if 'repos' not in tool_id:
+            return False
+        # It is a valid tool ID, and it is from a toolshed. Check if it's
+        # missing from the toolbox.
+        if trans.app.toolbox.get_tool( tool_id ) is None:
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: I'd change this to `self.app`  and drop `trans` as a parameter.

Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: Actually, I'd move this whole function to the toolbox and give it a signature like `is_missing_shed_tool(self, tool_id)`

**Comment by martenson** [comment]
#2769 is merged, so this is ready for review?

**Comment by davebx** [comment]
@martenson Yeah, it's all clear to test and critique.

**Comment by davebx** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
closing this just to test whether GH picks up jenkins' changes on reopen

**Comment by martenson** [comment]
Pushing this to 17.01, it needs some polishing and @davebx and me are on it.

**Review by martenson — CHANGES_REQUESTED** [review]
I am unable to preview or add to queue tools from workflow installed from the Gaius TS. Galaxy error below

```
"GET /api/tool_shed_repositories/shed_repository?tool_ids=gaius.bx.psu.edu%3A8008%2Frepos%2Fbgruening%2Fdeeptools%2Fdeeptools_computeGCBias%2F1.1.0%2Cgaius.bx.psu.edu%3A8008%2Frepos%2Fbgruening%2Fdeeptools%2Fdeeptools_computeMatrix%2F1.1.0%2Cgaius.bx.psu.edu%3A8008%2Frepos%2Fbgruening%2Fdeeptools%2Fdeeptools_correctGCBias%2F1.1.0%2Cgaius.bx.psu.edu%3A8008%2Frepos%2Fbgruening%2Fdeeptools%2Fdeeptools_heatmapper%2F1.1.0 HTTP/1.1" 500 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.95 Safari/537.36"
galaxy.web.framework.decorators ERROR 2017-01-13 10:29:47,078 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/Users/marten/devel/git/galaxy/lib/galaxy/web/framework/decorators.py", line 282, in decorator
    rval = func( self, trans, *args, **kwargs )
  File "/Users/marten/devel/git/galaxy/lib/galaxy/web/framework/decorators.py", line 98, in decorator
    return func( self, trans, *args, **kwargs )
  File "/Users/marten/devel/git/galaxy/lib/galaxy/webapps/galaxy/api/tool_shed_repositories.py", line 271, in shed_repository
    found_repository = json.loads( util.url_get( tool_shed_url, params=dict( tool_ids=','.join( tool_ids ) ), pathspec=[ 'api', 'repositories' ] ) )
  File "/Users/marten/devel/git/galaxy/lib/galaxy/util/__init__.py", line 1488, in url_get
    response = urlopener.open( full_url )
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 437, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 550, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 475, in error
    return self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 558, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 500: Internal Server Error
```

screenshot: 

screenshot 2017-01-13 10 31 43

**Comment by nsoranzo** [comment]
Superseded by #3626.

**Pull request closed without merging**